### PR TITLE
[AArch64] Support scalable offsets with isLegalAddressingMode

### PR DESCRIPTION
**Pull request by huntergr-arm**
Allows us to indicate that an addressing mode featuring a vscale-relative immediate offset is supported.

See the RFC for reference: https://discourse.llvm.org/t/rfc-vscale-aware-loopstrengthreduce/77131

### CODE REVIEW

**Comment by davemgreen** [comment]
This sounds like a nicer approach that adding the new type. Did you consider just having another `int64_t ScalableOffset` separate from the `Offset`? It might not really come up very often that an architecture has both at the same time, but it would be a little more general and fit into AddressingModes being a collection of things all added together.

**Review by davemgreen — COMMENTED** [review]
Thanks. I think this looks good, but had a couple of comments/suggestions.

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 1]
```diff
@@ -1530,7 +1530,8 @@ bool SITargetLowering::isLegalMUBUFAddressingMode(const AddrMode &AM) const {
 
 bool SITargetLowering::isLegalAddressingMode(const DataLayout &DL,
                                              const AddrMode &AM, Type *Ty,
-                                             unsigned AS, Instruction *I) const {
+                                             unsigned AS,
+                                             Instruction *I) const {
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:1534]: Unrelated formatting change spread through the targets

**Review by RKSimon — COMMENTED** [review]
How are targets supposed to know when the scalable offset is supposed to be used? Please can you add useful descriptions to the doxygen comments?

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 2]
```diff
@@ -16374,15 +16374,31 @@ bool AArch64TargetLowering::isLegalAddressingMode(const DataLayout &DL,
 
   if (Ty->isScalableTy()) {
     if (isa<ScalableVectorType>(Ty)) {
+      // See if we have a foldable vscale-based offset, for vector types which
+      // are either legal or smaller than the minimum; more work will be
+      // required if we need to consider addressing for types which need
+      // legalization by splitting.
+      uint64_t VecNumBytes = DL.getTypeSizeInBits(Ty).getKnownMinValue() / 8;
+      if (AM.HasBaseReg && !AM.BaseOffs && AM.ScalableOffset && !AM.Scale &&
+          (AM.ScalableOffset % VecNumBytes == 0) && VecNumBytes <= 16 &&
```
[the remark this answers is not in the document]
Comment (huntergr-arm) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp]: I check that the minimum number of bytes is <= 16 (so at most one legal SVE vector) and that it's a power-of-two. I would have restricted it to legal types only, but I've seen several loops which extended a smaller type; so a <vscale x 2 x i8> might be just fine as a type here to determine whether an addressing mode is legal, though it will be extended to <vscale x 2 x i64> in the actual register. We just care about the offset in memory, and from my reading of the ISA the `#imm, mul vl` addressing goes by the in-memory size.

I have indeed ignored larger-than-legal types for this initial work, as I don't have a motivating example where I need to improve addressing for them.

**Review by davemgreen — APPROVED** [review]
Thanks. This LGTM
